### PR TITLE
Set up test environment

### LINKED ISSUE #21: Set up test environment
## Why

To test the Beego application with automated testing

## Acceptance Criteria
- [x] Set up a test environment with `ginkgo` and `gomega`
- [x] Add test for registration

### PR DESCRIPTION
**Pull request by junan**
https://github.com/junan/google-scraper/projects/1#card-52861853

## What happened 👀

Set up test with [ginkgo](https://github.com/onsi/ginkgo) and [gomega](https://onsi.github.io/gomega/). Added registration backend test also.

## Insight 📝

- Set up test with [ginkgo](https://github.com/onsi/ginkgo) and [gomega](https://onsi.github.io/gomega/).
- Load `app.Conf` file manually in the DB initialization as this file does not load in the test
- Added controller, form, helpers. test
- Added `user_fabricator` to fabricate user in test
- Added some helper functions(flash message, http request) by mostly copying from the beego source code.
## Proof Of Work 📹

Spec should be passed https://github.com/junan/google-scraper/actions/runs/533520217

### CODE REVIEW

File: controllers/registration_test.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
+package controllers_test
+
+import (
+	"net/http"
+	"net/url"
+	"strings"
+	
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	. "google-scraper/tests/testing_helpers"
```
Comment (gutakk) [inline controllers/registration_test.go]: How about grouping the import?

File: controllers/registration_test.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+		It("returns 200 status code", func() {
+			response := MakeRequest("GET", "/register", nil)
+
+			Expect(response.Code).To(Equal(http.StatusOK))
+		})
+	})
+
+	Describe("POST", func() {
+		Context("given valid params", func() {
+			It("sets flash success message", func() {
```
Comment (gutakk) [inline controllers/registration_test.go:30]: Should we have redirect to correct path case?

File: controllers/registration_test.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+				body := strings.NewReader(registrationForm.Encode())
+				response := MakeRequest("POST", "/register", body)
+				flash := GetFlash(response.Result().Cookies())
+
+				Expect(flash.Data["success"]).To(Equal("Account has been created successfully"))
+			})
+		})
+
+		Context("given invalid params", func() {
+			It("sets flash error message", func() {
```
Comment (gutakk) [inline controllers/registration_test.go:45]: Should we have the render the correct page case?

File: forms/registration_form_test.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+import (
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	_ "google-scraper/initializers"
+	. "google-scraper/tests/fabricators"
+	. "google-scraper/tests/testing_helpers"
+
+	"google-scraper/forms"
```
Comment (gutakk) [inline forms/registration_form_test.go]:
```suggestion
	"google-scraper/forms"
```

File: forms/registration_form_test.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+				It("returns the correct error message and does NOT create new user", func() {
+					registrationForm := forms.RegistrationForm{
+						Name:     "John",
+						Email:    "invalid-email",
+						Password: "secret",
+					}
+					user, err := registrationForm.Save()
+
+					Expect(user).To(BeNil())
+					Expect(err.Error()).To(Equal("Email Must be a valid email address"))
```
Comment (gutakk) [inline forms/registration_form_test.go]: Can we uppercase only the first letter of the sentence so it should be `Email must be ...`? also another cases.

File: initializers/db.go [hunk 6]
```diff
@@ -12,6 +12,8 @@ import (
 )
 
 func init() {
+	LoadAppConfig()
+
 	dbUrl, _ := web.AppConfig.String("dbUrl")
```
Comment (gutakk) [inline initializers/db.go]: How about handle the error?

File: tests/testing_helpers/db.go [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+import (
+	"fmt"
+
+	"github.com/beego/beego/v2/client/orm"
+	"github.com/beego/beego/v2/core/logs"
+)
+
+func TruncateTable(tableName string) {
+	o := orm.NewOrm()
+	rqwSql := fmt.Sprintf("TRUNCATE TABLE \"%s\";", tableName)
```
Comment (gutakk) [inline tests/testing_helpers/db.go]: Should this be `rawSql`?

File: tests/testing_helpers/db.go [hunk 8]
```diff
@@ -0,0 +1,18 @@
+package test_helpers
```
Comment (gutakk) [inline tests/testing_helpers/db.go]: Should this be `package testing_helpers` to make it consistent with directory name?

File: controllers/registration_test.go [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+				form := url.Values{
+					"name":     {""},
+					"email":    {"john@example.com"},
+					"password": {"secret"},
+				}
+				body := strings.NewReader(form.Encode())
+				response := MakeRequest("POST", "/register", body)
+				flash := GetFlash(response.Result().Cookies())
+
+				Expect(flash.Data["error"]).To(Equal("Name Can not be empty"))
```
Comment (carryall) [inline controllers/registration_test.go]: @junan Can we set the default error message for validation so we can have a nicer error message `Name can not be empty`
```
validation.SetDefaultMessage(...)
```

File: forms/registration_form_test.go [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+					Password: "secret",
+				}
+				user, err := registrationForm.Save()
+
+				Expect(user.Id).NotTo(BeNil())
+				Expect(err).To(BeNil())
+			})
+		})
+
+		Context("given invalid attributes", func() {
```
Comment (carryall) [inline forms/registration_form_test.go:33]: @junan Should we add the case when the name is missing?

**Review by olivierobert — COMMENTED** [review]
Others have raised all issues. Make sure to clean up the PR title to "Set up Testing". There is no need for "chore/". This seems to have been taken from the branch name 🙈 

image

File: controllers/registration_test.go (hunk 2 above)
Comment (junan) [reply controllers/registration_test.go:30]: @gutakk I have not added the functionality to redirect to a new path after successful registration. Currently it stays same `/register` path. I will add it once I redirect to a new successful page.

File: controllers/registration_test.go (hunk 3 above)
Comment (junan) [reply controllers/registration_test.go:45]: I did not find for any render matcher. So I added a test case for not redirecting to any new page.

File: controllers/registration_test.go (hunk 2 above)
Comment (carryall) [reply controllers/registration_test.go:30]: @junan Can we add a `TODO` comment about this?

File: tests/fabricators/user_fabricator.go [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+
+import (
+	. "google-scraper/helpers"
+	. "google-scraper/models"
+
+	"github.com/beego/beego/v2/client/orm"
+	"github.com/beego/beego/v2/core/logs"
+)
+
+func FabricateUser(name string, email string, password string) {
```
Comment (carryall) [inline tests/fabricators/user_fabricator.go]: @junan How about return the created user also, it might be useful later 🙌🏼

**Review by carryall — COMMENTED** [review]
@junan I just noticed that there is no password confirmation on the registration form, should we also add it? 🤔

**Comment by junan** [comment]
> @junan I just noticed that there is no password confirmation on the registration form, should we also add it?
 
@carryall I intentionally did not add it to reduce extra step. It is also not mandatory and many sites(facebook,basecamp, shopify. etc) don,t add it these days.

File: controllers/registration_test.go (hunk 3 above)
Comment (gutakk) [reply controllers/registration_test.go:44]: Maybe you can check Hoang's code
https://github.com/hoangmirs/go-scraper/pull/26/commits/a2a6d97fab6833383acc9a64404720f6a70fc111

**Review by gutakk — APPROVED** [review]

**Review by olivierobert — APPROVED** [review]

**Pull request merged**